add more detailed documentation for sso in readthedocs (#30282)
dimagi/commcare-hq: docs/sso/architecture.rst

Architecture
==============

.. note::

>     Everything related on SSO on CommCare HQ can be found in `corehq.apps.sso`

Reviewer: ```suggestion
    Everything related to SSO on CommCare HQ can be found in `corehq.apps.sso`
```

of their username. An **email domain** is the portion of an email address that
follows the `@` sign.

We tie email domains to `IdentityProviders` with the `AuthenticatedEmailDomain`
model.

If a user's email domain matches an `AuthenticatedEmailDomain`, during the
SSO login process they will be directed to the login workflow determined by
the active `IdentityProvider` associated with the `AuthenticatedEmailDomain`.

.. note::
    A user will only be forced to use SSO at login and sign up if
    `ENFORCE_SSO_LOGIN` in `localsettings` is set to `True`. Otherwise, they
    will be able to login with a username and password (if they created an
    account originally this way) or by visiting the `IdentityProvider`'s
    login URL (see the `get_login_url` method).

UserExemptFromSingleSignOn
--------------------------

Even if the `ENFORCE_SSO_LOGIN` flag in `localsettings` is set to `True`, we
still need to allow certain users the ability to always login to CommCare HQ
with their username and password. Generally, these users are one or two
enterprise admins (or other special users).

We require at least one user to be exempt from SSO login in the event that

> the `IdentityProvider` is mis-configured or the certificate expires and a user

Reviewer: ```suggestion
the `IdentityProvider` is misconfigured or the certificate expires and a user
```

needs to gain access to their enterprise console to fix the situation.

TrustedIdentityProvider
-----------------------

Project spaces that are not associated with the `BillingAccount` tied to a given
`IdentityProvider` **do not** automatically trust users who have authenticated
with that `IdentityProvider`.

In order for a user to access a project space that is outside of their
`IdentityProvider`'s jurisdiction, an admin of that project space must first
agree to trust the `IdentityProvider` associated with that user.

This trust can either be established from the Web Users list or when inviting
a web user that is using SSO to the project.

> Once a trust is established, any user authenticating with that `IdentityProvider`

Reviewer: ```suggestion
Once trust is established, any user authenticating with that `IdentityProvider`
```
I could go either way here 🤷
